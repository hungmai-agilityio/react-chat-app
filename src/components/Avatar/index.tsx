import { clsx } from 'clsx';

// Constants
import { SIZE } from '@/constants';

interface AvatarProps {
  avatar?: string;
  name?: string;
  size?: SIZE.SMALL | SIZE.LARGE;
  styles?: string;
  circle?: boolean;
  onClick?: () => void;
}

const Avatar = ({
  avatar,
  name,
  size,
  styles,
  circle,
  onClick
}: AvatarProps) => {
  const nameAvatar = name ? name.charAt(0).toUpperCase() : '';

  const sizeClasses = {
    [SIZE.SMALL]: 'w-avatar-sm h-avatar-sm text-lg',
    [SIZE.LARGE]: 'w-avatar-lg h-avatar-lg text-8xl'
  };

  const sizeClass = sizeClasses[size || SIZE.SMALL];

  if (!avatar) {
    return (
      <div
        data-testid="avatar"
        onClick={onClick}
        className={clsx(
          'text-primary bg-gray-300 flex items-center justify-center font-semibold',
          circle && 'rounded-full',
          styles,
          sizeClass
        )}
      >
        {nameAvatar}
      </div>
    );
  }

  return (
    <img
      data-testid="avatar"
      onClick={onClick}
      src={avatar}
      alt={name}
      className={clsx(
        'object-cover font-bold',
        circle && 'rounded-full',
        styles,
        sizeClass
      )}
    />
  );
};

export default Avatar;
