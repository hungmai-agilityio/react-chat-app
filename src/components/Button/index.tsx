import { clsx } from 'clsx';
import { memo } from 'react';

// Constants
import { SIZE, TYPE } from '@/constants';

// FontAwesome
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface ButtonProps {
  name: string;
  size?: SIZE.MINI | SIZE.SMALL | SIZE.MEDIUM | SIZE.LARGE;
  variant?: TYPE.PRIMARY | TYPE.SECOND | TYPE.TERTIARY;
  onClick: (e: React.FormEvent<HTMLButtonElement>) => void;
  styles?: string;
  isDisabled?: boolean;
  iconLeft?: IconDefinition;
  iconRight?: IconDefinition;
}

const Button = ({
  name,
  size,
  variant,
  onClick,
  styles,
  isDisabled,
  iconLeft,
  iconRight
}: ButtonProps) => {
  const sizeClasses = {
    [SIZE.MINI]: 'w-24 h-9',
    [SIZE.SMALL]: 'w-32 h-10',
    [SIZE.MEDIUM]: 'w-40 h-9',
    [SIZE.LARGE]: 'w-full h-10'
  };

  const typeClasses = {
    [TYPE.PRIMARY]: 'bg-primary text-white',
    [TYPE.SECOND]: 'bg-tertiary text-primary',
    [TYPE.TERTIARY]: 'hover:bg-tertiary text-primary'
  };

  const sizeClass = sizeClasses[size || SIZE.SMALL];
  const typeClass = typeClasses[variant || TYPE.PRIMARY];

  return (
    <button
      className={clsx(
        'font-semibold hover:brightness-90',
        isDisabled &&
          'bg-tertiary cursor-not-allowed border hover:filter-none text-stone-500',
        sizeClass,
        typeClass,
        styles
      )}
      onClick={onClick}
      disabled={isDisabled}
    >
      {iconLeft && <FontAwesomeIcon icon={iconLeft} className="mr-2" />}
      {name}
      {iconRight && <FontAwesomeIcon icon={iconRight} className="ml-2" />}
    </button>
  );
};

const MemoizedButton = memo(Button);
export default MemoizedButton;
