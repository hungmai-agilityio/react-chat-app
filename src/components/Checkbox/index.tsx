import { clsx } from 'clsx';
import { ReactNode } from 'react';

interface CheckboxProps {
  id: string;
  children: ReactNode;
  isChecked: boolean;
  onChange: (id: string) => void;
}

const Checkbox = ({ id, children, isChecked, onChange }: CheckboxProps) => {
  // Handle checkbox
  const handleChecked = () => {
    onChange(id);
  };

  return (
    <div
      id={id}
      className={clsx('pl-3 flex items-center gap-4 w-full', {
        'bg-blue-50': isChecked
      })}
    >
      <input
        type="checkbox"
        checked={isChecked}
        className="w-5 h-5 accent-primary"
        onChange={handleChecked}
      />
      <label onClick={handleChecked} className="w-full">
        {children}
      </label>
    </div>
  );
};

export default Checkbox;
