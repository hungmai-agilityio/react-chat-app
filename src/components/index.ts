// Exports default
export { default as Avatar } from '@/components/Avatar';
export { default as Button } from '@/components/Button';
export { default as Divider } from '@/components/Divider';
export { default as Dropdown } from '@/components/Dropdown';
export { default as UserMenu } from '@/components/Dropdown/UserMenu';
export { default as ErrorBoundary } from '@/components/ErrorBoundary';
export { default as Form } from '@/components/Form';
export { default as SignIn } from '@/components/Form/SignIn';
export { default as SignUp } from '@/components/Form/SignUp';
export { default as Input } from '@/components/Input';
export { default as Loading } from '@/components/Loading';
export { default as Spinner } from '@/components/Loading/Spinner';
export { default as ChatMessage } from '@/components/Chat';

export { default as Modal } from '@/components/Modal';
export { default as ModalAction } from '@/components/Modal/Action';
export { default as AddMember } from '@/components/Modal/AddMember';
export { default as ModalInfo } from '@/components/Modal/Info';
export { default as Profile } from '@/components/Modal/Profile';
export { default as Search } from '@/components/Search';
export { default as Upload } from '@/components/Upload';
export { default as ListUser } from '@/components/UserRoom/List';

// Exports named
export * from '@/components/Checkbox';
export * from '@/components/Dropdown';
export * from '@/components/Chip';
export * from '@/components/Item';
export * from '@/components/Message';
export * from '@/components/Message/GroupMessage';
export * from '@/components/Tabs';
export * from '@/components/Tabs/Tab';
export * from '@/components/Tabs/Panel';
export * from '@/components/UserRoom';
