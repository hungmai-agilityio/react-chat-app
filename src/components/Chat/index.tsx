import {
  ChangeEvent,
  Dispatch,
  SetStateAction,
  useCallback,
  useEffect,
  useRef,
  useState
} from 'react';
import { v4 as uuidv4 } from 'uuid';
import clsx from 'clsx';

// Hooks
import { useOutsideClick } from '@/hooks';

// Components
import { Button, Input } from '@/components';

// Interfaces
import { IMessage, IChat } from '@/interfaces';

// Services
import { createChat, sendMessage, updateMessage } from '@/services';

// Constants
import { SIZE, TYPE } from '@/constants';

interface ChatMessageProps {
  currentUserId: string;
  setMessages: Dispatch<SetStateAction<IMessage[]>>;
  isDisabled: boolean;
  chatData?: IChat | null;
  selectedUser?: string;
  chats?: IChat[];
  handleCancelEdit: () => void;
  isEdit: boolean;
  editMessage?: IMessage | null;
}

const ChatMessage = ({
  currentUserId,
  setMessages,
  isDisabled,
  chatData,
  selectedUser,
  chats,
  handleCancelEdit,
  isEdit,
  editMessage
}: ChatMessageProps) => {
  const [value, setValue] = useState<string>('');
  const [isFocused, setIsFocused] = useState<boolean>(false);

  const inputRef = useRef<HTMLDivElement>(null);

  // Hook to handle outside click
  useOutsideClick({
    ref: inputRef,
    handler: () => setIsFocused(false)
  });

  // Initialize value if editing
  useEffect(() => {
    if (isEdit && editMessage) {
      setValue(editMessage.message);
    }
  }, [isEdit, editMessage]);

  // Handle change value message
  const handleChangeValue = (
    event: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    setValue(event.target.value);
  };

  // Handle send message
  const handleSend = useCallback(async () => {
    if (!value.trim().length) return;

    let roomId = chatData?.id;

    if (!roomId && selectedUser) {
      const existingChat = chats?.find(
        (chat) =>
          !chat.isGroup &&
          chat.members.includes(currentUserId) &&
          chat.members.includes(selectedUser)
      );

      if (!existingChat) {
        const newChat: IChat = {
          id: uuidv4(),
          title: '',
          avatar: '',
          members: [currentUserId, selectedUser],
          isGroup: false
        };

        const createChatResponse = await createChat(newChat);
        roomId = createChatResponse.data?.id;
      }

      if (existingChat) {
        roomId = existingChat?.id;
      }
    }

    const newMessage: IMessage = {
      id: uuidv4(),
      message: value,
      sender: currentUserId,
      time_stamp: new Date(),
      roomId: roomId || ''
    };

    await sendMessage(newMessage);

    setMessages((prevMessages: IMessage[]) => {
      const messageExists = prevMessages.some(
        (msg) => msg.id === newMessage.id
      );
      if (messageExists) return prevMessages;

      return [...prevMessages, newMessage];
    });
    setValue('');
  }, [value, currentUserId, chatData, selectedUser, chats]);

  // Handle update message
  const handleUpdateMessage = useCallback(async () => {
    if (!value.trim().length || !editMessage) return;

    const updatedMessage: Partial<IMessage> = {
      id: editMessage.id,
      message: value,
      sender: currentUserId,
      isEdit: true
    };

    const { data } = await updateMessage(updatedMessage as IMessage);

    if (data) {
      setMessages((prev) =>
        prev.map((msg) =>
          msg.id === editMessage.id
            ? { ...msg, message: value, isEdit: true }
            : msg
        )
      );
    }

    setValue('');
    handleCancelEdit();
  }, [value, editMessage, currentUserId, handleCancelEdit]);

  // Handle press enter to send message
  const handleKeyDown = (
    e: React.KeyboardEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    if (e.key === 'Enter' && !e.shiftKey) {
      e.preventDefault();
      if (isEdit) {
        handleUpdateMessage();
      } else {
        handleSend();
      }
    }
  };

  return (
    <div
      ref={inputRef}
      onFocus={() => setIsFocused(true)}
      className={clsx('relative p-2 transition-all')}
    >
      <Input
        htmlFor="message-value"
        onChange={handleChangeValue}
        name="message"
        value={value}
        variant={TYPE.SECOND}
        disabled={isDisabled}
        onKeyDown={handleKeyDown}
      />
      {isFocused && (
        <div className="float-right">
          <div className="flex gap-5">
            {isEdit && (
              <Button
                name="Cancel"
                onClick={handleCancelEdit}
                size={SIZE.MINI}
              />
            )}
            <Button
              name={isEdit ? 'Update' : 'Send'}
              onClick={isEdit ? handleUpdateMessage : handleSend}
              disabled={value.length === 0 || isDisabled}
              size={SIZE.MINI}
            />
          </div>
        </div>
      )}
    </div>
  );
};

export default ChatMessage;
