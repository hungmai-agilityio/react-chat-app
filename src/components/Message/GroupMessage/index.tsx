import { memo, useMemo } from 'react';

// Interfaces
import { IMessage } from '@/interfaces';

// Components
import Divider from '@/components/Divider';
import Chip from '@/components/Chip';
import { Message } from '@/components/Message';

interface GroupedMessagesProps {
  messages: IMessage[];
  userProfiles: Map<string, { name: string; avatar: string }>;
  currentUser: string;
  onEdit?: (item: IMessage) => void;
  onDelete?: (id: string) => void;
}

export const GroupedMessages = memo(
  ({
    messages,
    userProfiles,
    currentUser,
    onEdit,
    onDelete
  }: GroupedMessagesProps) => {
    const handleEdit = (item: IMessage) => {
      if (onEdit) {
        onEdit(item);
      }
    };

    const handleRemove = (id: string) => {
      if (onDelete) {
        onDelete(id);
      }
    };
    // render and group messages by date
    const groupedMessages = useMemo(() => {
      if (!messages.length) return [];

      const grouped: { [key: string]: IMessage[] } = {};

      messages.forEach((message) => {
        const date = new Date(message.time_stamp);
        const dateString = date.toDateString();

        if (!grouped[dateString]) {
          grouped[dateString] = [];
        }
        grouped[dateString].push(message);
      });

      return Object.keys(grouped).map((date) => ({
        date,
        messages: grouped[date]
      }));
    }, [messages]);

    return (
      <>
        {groupedMessages.map((group) => (
          <div key={group.date}>
            <Divider>
              <Chip>
                <p>{group.date}</p>
              </Chip>
            </Divider>
            {group.messages.map((item) => {
              const { name, avatar } = userProfiles.get(item.sender) || {
                name: '',
                avatar: ''
              };

              return (
                <Message
                  key={item.id}
                  item={item}
                  isCurrentUser={item.sender === currentUser}
                  avatar={avatar}
                  name={name}
                  isActive
                  onEdit={handleEdit}
                  onDelete={handleRemove}
                />
              );
            })}
          </div>
        ))}
      </>
    );
  }
);
