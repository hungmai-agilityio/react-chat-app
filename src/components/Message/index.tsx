import { clsx } from 'clsx';
import { memo } from 'react';

// FontAwesome
import {
  faCircle,
  faEllipsisVertical
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

// Constants
import { MESSAGE_OPTIONS, POSITION } from '@/constants';

// Interfaces
import { IMessage } from '@/interfaces';

// Utils
import { timeStampConversion } from '@/utils';

// Components
import Avatar from '@/components/Avatar';
import Dropdown, { DropdownItem } from '@/components/Dropdown';

interface MessageProps {
  item: IMessage;
  isCurrentUser: boolean;
  avatar: string;
  name: string;
  isActive?: boolean;
  onEdit?: (item: IMessage) => void;
  onDelete?: (id: string) => void;
}

export const Message = memo(
  ({
    item,
    isCurrentUser,
    avatar,
    name,
    isActive,
    onEdit,
    onDelete
  }: MessageProps) => {
    /**
     * Handles selection of a dropdown item.
     * @param {DropdownItem} selectedItem
     */
    const handleActionSelect = (selectedItem: DropdownItem) => {
      switch (selectedItem.value) {
        case 'editMess':
          onEdit?.(item);
          break;
        case 'deleteMess':
          onDelete?.(item.id);
          break;
        default:
          break;
      }
    };

    return (
      <div
        className={clsx(
          'relative flex mb-5',
          isCurrentUser ? 'justify-end' : 'justify-start'
        )}
      >
        <div
          className={clsx('flex gap-3 max-w-message', {
            'flex-row-reverse': isCurrentUser
          })}
        >
          <div className="relative w-full max-w-avatar">
            <Avatar avatar={avatar} name={name} circle />
            <FontAwesomeIcon
              icon={faCircle}
              className={clsx(
                'absolute text-gray-400 border-xs border-white rounded-full top-6 left-7',
                { 'text-green-500': isActive }
              )}
            />
          </div>
          <div className="w-full">
            <div
              className={clsx('flex items-baseline gap-2', {
                'justify-end': isCurrentUser
              })}
            >
              <p>{name}</p>
              <span className="text-xs text-gray-400">
                {timeStampConversion(item.time_stamp)}
              </span>
            </div>

            <div
              className={clsx(
                'p-2 text-dark mt-2 group pr-7 max-w-max',
                isCurrentUser
                  ? 'bg-second rounded-mess-user float-right'
                  : 'bg-tertiary rounded-mess-member'
              )}
            >
              {item.message}{' '}
              {item.isEdit && (
                <span className="text-sm text-gray-500">(edited)</span>
              )}
            </div>

            {isCurrentUser && (
              <div className="absolute top-6 right-7">
                <Dropdown
                  icon={faEllipsisVertical}
                  items={MESSAGE_OPTIONS}
                  onSelect={handleActionSelect}
                  position={POSITION.TOP_RIGHT}
                  isHover={false}
                />
              </div>
            )}
          </div>
        </div>
      </div>
    );
  }
);
