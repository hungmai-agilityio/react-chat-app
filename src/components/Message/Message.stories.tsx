import { Meta, StoryObj } from '@storybook/react';

// Components
import { Message } from '@/components/Message';
import { mockMessage, mockMessageEdit } from '@/mocks';

const meta: Meta<typeof Message> = {
  component: Message,
  tags: ['autodocs'],
  argTypes: {
    item: {
      description: 'Message data'
    },
    isCurrentUser: {
      description: 'Check if the user is the one currently logged in'
    },
    avatar: {
      description: 'Render image for img tags'
    },
    name: {
      description: 'Set name if there is no avatar or the avatar is corrupted'
    },
    isActive: {
      description: 'Check if the user is connected?'
    },
    onEdit: {
      description: 'Handles message editing'
    },
    onDelete: {
      description: 'Handle message deletion'
    }
  }
};

export default meta;

type Story = StoryObj<typeof Message>;

export const UserMessage: Story = {
  args: {
    item: mockMessageEdit,
    isCurrentUser: true,
    avatar: '',
    name: 'User Name',
    isActive: false,
    onEdit: () => console.log('Edit message'),
    onDelete: () => console.log('Delete message')
  }
};

export const MemberMessage: Story = {
  args: {
    item: mockMessage,
    isCurrentUser: false,
    avatar: '',
    isActive: true,
    name: 'Member Name'
  }
};
