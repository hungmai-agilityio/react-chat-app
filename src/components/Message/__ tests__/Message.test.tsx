import { fireEvent, render, screen } from '@testing-library/react';

// Mock data
import { mockMessage } from '@/mocks';

// Components
import { Message } from '@/components/Message';

describe('Message Component', () => {
  const defaultProps = {
    item: mockMessage,
    avatar: '',
    name: 'John Doe',
    isActive: true
  };

  it('renders UI for a member', () => {
    const { container } = render(
      <Message {...defaultProps} isCurrentUser={false} />
    );

    expect(screen.getByText('Hello')).toBeInTheDocument();
    expect(screen.getByText('John Doe')).toBeInTheDocument();
    expect(screen.getByText('24.04.2024')).toBeInTheDocument();

    const iconDropdown = container.querySelector('.fa-ellipsis-vertical');
    expect(iconDropdown).toBeNull();
  });

  it('renders UI for a user with dropdown', () => {
    const { container } = render(
      <Message {...defaultProps} isCurrentUser={true} />
    );

    expect(screen.getByText('Hello')).toBeInTheDocument();
    expect(screen.getByText('John Doe')).toBeInTheDocument();
    expect(screen.getByText('24.04.2024')).toBeInTheDocument();

    const iconDropdown = container.querySelector('.fa-ellipsis-vertical');
    expect(iconDropdown).toBeInTheDocument();
  });

  it('opens dropdown and displays options when clicked', () => {
    const { container } = render(
      <Message {...defaultProps} isCurrentUser={true} />
    );
    const iconDropdown = container.querySelector('.fa-ellipsis-vertical');

    fireEvent.click(iconDropdown!);

    expect(screen.getByText('Edit')).toBeInTheDocument();
    expect(screen.getByText('Delete')).toBeInTheDocument();
  });

  test('triggers onEdit when "Edit" is selected from dropdown', () => {
    const onEdit = jest.fn();
    const { container } = render(
      <Message {...defaultProps} isCurrentUser={true} onEdit={onEdit} />
    );

    const iconDropdown = container.querySelector('.fa-ellipsis-vertical');
    fireEvent.click(iconDropdown!);

    fireEvent.click(screen.getByText('Edit'));

    expect(onEdit).toHaveBeenCalledTimes(1);
  });

  test('triggers onDelete when "Delete" is selected from dropdown', () => {
    const onDelete = jest.fn();
    const { container } = render(
      <Message {...defaultProps} isCurrentUser={true} onDelete={onDelete} />
    );

    const iconDropdown = container.querySelector('.fa-ellipsis-vertical');
    fireEvent.click(iconDropdown!);

    fireEvent.click(screen.getByText('Delete'));

    expect(onDelete).toHaveBeenCalledTimes(1);
  });

  it('should be render match to snapshot Member', () => {
    const { container } = render(
      <Message {...defaultProps} isCurrentUser={false} />
    );

    expect(container).toMatchSnapshot();
  });

  it('should be render match to snapshot user', () => {
    const { container } = render(
      <Message {...defaultProps} isCurrentUser={true} />
    );

    expect(container).toMatchSnapshot();
  });
});
