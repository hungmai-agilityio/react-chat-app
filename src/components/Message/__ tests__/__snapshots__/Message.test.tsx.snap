// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message Component should be render match to snapshot Member 1`] = `
<div>
  <div
    class="relative flex mb-5 justify-start"
  >
    <div
      class="flex gap-3 max-w-message"
    >
      <div
        class="relative w-full max-w-avatar"
      >
        <div
          class="text-primary bg-gray-300 flex items-center justify-center font-semibold rounded-full w-avatar-sm h-avatar-sm text-lg"
          data-testid="avatar"
        >
          J
        </div>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle absolute text-gray-400 border-xs border-white rounded-full top-6 left-7 text-green-500"
          data-icon="circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512z"
            fill="currentColor"
          />
        </svg>
      </div>
      <div
        class="w-full"
      >
        <div
          class="flex items-baseline gap-2"
        >
          <p>
            John Doe
          </p>
          <span
            class="text-xs text-gray-400"
          >
            24.04.2024
          </span>
        </div>
        <div
          class="p-2 text-dark mt-2 group pr-7 max-w-max bg-tertiary rounded-mess-member"
        >
          Hello
           
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Message Component should be render match to snapshot user 1`] = `
<div>
  <div
    class="relative flex mb-5 justify-end"
  >
    <div
      class="flex gap-3 max-w-message flex-row-reverse"
    >
      <div
        class="relative w-full max-w-avatar"
      >
        <div
          class="text-primary bg-gray-300 flex items-center justify-center font-semibold rounded-full w-avatar-sm h-avatar-sm text-lg"
          data-testid="avatar"
        >
          J
        </div>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle absolute text-gray-400 border-xs border-white rounded-full top-6 left-7 text-green-500"
          data-icon="circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512z"
            fill="currentColor"
          />
        </svg>
      </div>
      <div
        class="w-full"
      >
        <div
          class="flex items-baseline gap-2 justify-end"
        >
          <p>
            John Doe
          </p>
          <span
            class="text-xs text-gray-400"
          >
            24.04.2024
          </span>
        </div>
        <div
          class="p-2 text-dark mt-2 group pr-7 max-w-max bg-second rounded-mess-user float-right"
        >
          Hello
           
        </div>
        <div
          class="absolute top-6 right-7"
        >
          <div
            class="relative"
          >
            <div
              class="flex items-center cursor-pointer px-4"
            >
              <div
                class="flex items-center justify-center cursor-pointer text-gray-400 h-10 w-10 rounded-full"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-ellipsis-vertical "
                  data-icon="ellipsis-vertical"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 128 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M64 360a56 56 0 1 0 0 112 56 56 0 1 0 0-112zm0-160a56 56 0 1 0 0 112 56 56 0 1 0 0-112zM120 96A56 56 0 1 0 8 96a56 56 0 1 0 112 0z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
